tests/api/test_metrics_steps.py

```diff
@@ -0,0 +1,407 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import time
+
+import pytest
+from snowflake.cli.api.metrics import (
+    CLIMetrics,
+    CLIMetricsInvalidUsageError,
+    _CLIMetricsStep,
+)
+
+
+def test_metrics_steps_initialization_empty():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+
+    # then
+    assert metrics.steps == []
+
+
+def test_metrics_steps_time_is_valid():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("step1"):
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.NAME_KEY] == "step1"
+
+
+def test_metrics_steps_name_is_valid():
+    # given
+    metrics = CLIMetrics()
+    time_before_step = time.time()
+
+    # when
+    with metrics.track_step("step1"):
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.START_TIME_KEY] >= time_before_step
+    assert step1[_CLIMetricsStep.EXECUTION_TIME_KEY] > 0
+
+
+def test_metrics_steps_error_caught():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("step1"):
+        try:
+            raise RuntimeError()
+        except RuntimeError:
+            pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.ERROR_KEY] is None
+
+
+def test_metrics_steps_error_uncaught():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    try:
+        with metrics.track_step("step1"):
+            raise RuntimeError()
+    except RuntimeError:
+        pass
```

Comment:
never do this in tests. assert that the exception is indeed propagated instead. Pytest makes that trivial.